src/infra/core/src/remote_alias_resolver_impl.rs

```diff
@@ -0,0 +1,296 @@
+    async fn resolve_remote_alias(
+        &self,
+        local_dataset_handle: &odf::DatasetHandle,
+        dataset_push_target_maybe: Option<odf::DatasetPushTarget>,
+        remote_alias_kind: RemoteAliasKind,
+    ) -> Result<RemoteAliasRef, ResolveAliasError> {
+        let repo_name: odf::RepoName;
+        let mut account_name = None;
+        let mut dataset_name = None;
+
+        if let Some(dataset_push_target) = &dataset_push_target_maybe {
+            match dataset_push_target {
+                odf::DatasetPushTarget::Alias(dataset_alias_remote) => {
+                    repo_name = dataset_alias_remote.repo_name.clone();
+                    account_name.clone_from(&dataset_alias_remote.account_name);
+                    dataset_name = Some(dataset_alias_remote.dataset_name.clone());
+                }
+                odf::DatasetPushTarget::Url(url_ref) => {
+                    return Ok(RemoteAliasRef::new(
+                        url_ref.clone(),
+                        None,
+                        dataset_name,
+                        account_name,
+                    ));
+                }
+                odf::DatasetPushTarget::Repository(repository_name) => {
+                    repo_name = repository_name.clone();
+                }
+            }
+        } else {
+            if let Some(remote_url) = self
+                .fetch_remote_url(local_dataset_handle, remote_alias_kind)
+                .await?
+            {
+                return Ok(RemoteAliasRef::new(
+                    remote_url,
+                    None,
+                    dataset_name,
+                    account_name,
+                ));
+            }
+            let remote_repo_names: Vec<_> = self.remote_repo_reg.get_all_repositories().collect();
+            if remote_repo_names.len() > 1 {
+                return Err(ResolveAliasError::AmbiguousRepository);
+            }
+            if let Some(repository_name) = remote_repo_names.first() {
+                repo_name = repository_name.clone();
+            } else {
+                return Err(ResolveAliasError::EmptyRepositoryList);
+            }
+        }
+        let remote_repo = self.remote_repo_reg.get_repository(&repo_name).int_err()?;
+
+        let access_token_maybe = self
+            .access_token_resolver
+            .resolve_odf_dataset_access_token(&remote_repo.url);
+        if account_name.is_none() {
+            account_name = RemoteAliasResolverApiHelper::resolve_remote_account_name(
+                &remote_repo.url,
+                access_token_maybe.as_ref(),
+            )
+            .await
+            .int_err()?;
+        }
+        let transfer_dataset_name = dataset_name.clone().unwrap_or(
+            self.resolve_remote_dataset_name(
+                local_dataset_handle,
+                &remote_repo.url,
+                access_token_maybe.as_ref(),
+            )
+            .await?,
+        );
```

Comment:
Isn't `resolve_remote_dataset_name` an expensive function? Written like this we will always be calling it even if `dataset_name` was not `None`.

Reply:
Yeah, it is very expensive, but it was my idea, if the user does not provide dataset_name I am trying to fetch it from a remote server.